gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java

```diff
@@ -0,0 +1,205 @@
+import com.google.protobuf.ByteString;
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+
+/** Implements WritableByteChannel to provide write access to GCS via java-storage client */
+class GoogleCloudStorageClientWriteChannel extends AbstractGoogleAsyncWriteChannel<Boolean> {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private final StorageResourceId resourceId;
+  private WriteChannel writeChannel;
+  private Boolean uploadSucceeded = false;
+  // TODO: not supported as of now
+  // private final String requesterPaysProject;
+
+  public GoogleCloudStorageClientWriteChannel(
+      Storage storage,
+      GoogleCloudStorageOptions storageOptions,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      ExecutorService uploadThreadPool) {
+    super(uploadThreadPool, storageOptions.getWriteChannelOptions());
+    this.resourceId = resourceId;
+    this.writeChannel = getClientWriteChannel(storage, resourceId, createOptions, storageOptions);
+  }
+
+  @Override
+  public void startUpload(InputStream pipeSource) throws IOException {
+    // Given that the two ends of the pipe must operate asynchronous relative
+    // to each other, we need to start the upload operation on a separate thread.
+    try {
+      uploadOperation = threadPool.submit(new UploadOperation(pipeSource, this.resourceId));
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Failed to start upload for '%s'", resourceId), e);
+    }
+  }
+
+  private static BlobInfo getBlobInfo(
+      StorageResourceId resourceId, CreateObjectOptions createOptions) {
+    BlobId blobId =
+        BlobId.of(
+            resourceId.getBucketName(), resourceId.getObjectName(), resourceId.getGenerationId());
+    BlobInfo blobInfo =
+        BlobInfo.newBuilder(blobId)
+            .setContentType(createOptions.getContentType())
+            .setContentEncoding(createOptions.getContentEncoding())
+            .setMetadata(encodeMetadata(createOptions.getMetadata()))
+            .build();
+    return blobInfo;
+  }
+
+  private static WriteChannel getClientWriteChannel(
+      Storage storage,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      GoogleCloudStorageOptions storageOptions) {
+    AsyncWriteChannelOptions channelOptions = storageOptions.getWriteChannelOptions();
+    WriteChannel writeChannel =
+        storage.writer(
+            getBlobInfo(resourceId, createOptions),
+            generateWriteOptions(createOptions, channelOptions));
+    writeChannel.setChunkSize(getUploadChunkSize(channelOptions.getNumberOfBufferedRequests()));
+
+    return writeChannel;
+  }
+
+  private static int getUploadChunkSize(int bufferedRequestCount) {
```

Comment:
I think that it will be better if we use `getUploadChunkSize` from options class directly: https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/babcabd1f922d65852c7f9e288b2949844f30524/util/src/main/java/com/google/cloud/hadoop/util/AsyncWriteChannelOptions.java#L71

Reply:
Done.